skywalking/agent/__init__.py

```diff
@@ -65,14 +76,40 @@ def __init():
         __protocol = KafkaProtocol()
 
     plugins.install()
+    __init_threading()
 
 
 def __fini():
     __protocol.report(__queue, False)
     __queue.join()
+    __finished.set()
+
+
+def __fork_before():
+    if config.protocol != 'http':
+        logger.warning('fork() not currently supported with %s protocol' % config.protocol)
```

Comment:
Is it possible to make gRPC work in fork? Like I said in DM, recreate a brand new gRPC channel in forked process?

Reply:
Tried closing down channel and recreating in both parent and child after fork. It is possible I did not do it right since I am not grpc expert but I got one of two results:
1. Worked in exactly one of the forks, parent or child, but not both.
2. Didn't work in either.

Reply:
I didn't mean to close parent channel and create in child process by reusing the agent in parent. What I meant is to start another independent agent in child process and leave the parent one there because there may be other things that may need to be traced in parent process. Can you take a look at 

https://github.com/apache/skywalking-python/blob/c73398544694ff7c64b083a5ee0d6d82578971cf/skywalking/trace/ipc/process.py#L30-L34

... and see whether that helps, it is generally what I propose to do in forked processes?

Reply:
I your current implementation, when new processes are spawned, the agent in parent process takes no effect then, right?

Reply:
> I didn't mean to close parent channel and create in child process by reusing the agent in parent. What I meant is to start another independent agent in child process and leave the parent one there because there may be other things that may need to be traced in parent process. Can you take a look at

I tried several things like:
* Not doing anything before the fork then creating new `GrpcServiceManagementClient` and `GrpcTraceSegmentReportService` in child.
* The above but closing channel in child before creating new.
* Closing the channel before fork then recreating in both parent and child.
* Instead of `close()`, use `unsubscribe()`.
* Both `unsubscribe()` then `close()` before fork or after in child.
* I did also try waiting for empty queue before allowing fork to proceed but that was unnecessary as I wasn't even sending anything before the fork, just for form.

I also forgot to mention there was a third result I was getting sometime, deadlock hang. It is possible I missed some permutations or a different function to call, but in general researching python grpc with multiprocessing on the the net I found basically the following answers, either 1. "don't do it", or 2. "grpc must be started only after forking everything, then maybe it will work". Here are some links:

https://github.com/googleapis/synthtool/issues/902
https://stackoverflow.com/questions/62798507/why-multiprocess-python-grpc-server-do-not-work

So as I said, it may be possible but I have not hit on how to do it. If you want to give it a shot I will add a simple test scrip to the end of this message. I also didn't test anything with Kafka and assume it will not work correctly forking until someone validates that.

As for current higher level flow, keep in mind it can be modified in the future according to what protocol is in use, but for now - Nothing special is done before fork or after in the parent. In those cases all threads and sockets and locks continue operating as if nothing had happened. In the child, new report and heartbeat threads are started since threads don't survive into children. And specifically in the http protocol, the duplicated sockets are closed and new ones are opened on next heartbeat or report.

There is a potential problem with the `__queue` object as a thread may have been holding an internal lock on it before fork and since that thread is no longer present the queue will remain in a locked state. Not sure how to resolve this yet, but it should be a very rare event. Even rarer may be the same lock problem with the `__finished` event, but I wouldn't expect that to happen basically ever.

Right now I have other stuff on my plate but if you have any suggestions on what to try I may revisit this at some point in the future. Or if you ant to try yourself, here is a test script:
```py
import multiprocessing as mp
import time

from skywalking.trace.context import get_context
from skywalking import agent, config

config.init(collector='127.0.0.1:11800', service='your awesome service')
agent.start()

def foo():
    with get_context().new_local_span('child before error'):
        pass

    time.sleep(2)  # this needed to flush send because python doesn't run atexit handlers on exit in forked children

    # import atexit
    # atexit._run_exitfuncs()

if __name__ == '__main__':
    p = mp.Process(target = foo, args = ())

    with get_context().new_local_span('parent before start'):
        pass

    p.start()

    time.sleep(1)

    with get_context().new_local_span('parent after start'):
        pass

    p.join()

    with get_context().new_local_span('parent after join'):
        pass
```

Reply:
Just run the test script for several times, all 4 spans are correctly reported

Reply:
With grpc? Really?

Reply:
Can you give me version numbers of everything relevant, python included?

Reply:
If you got this far, you should check with some spans before the fork as well to see what happens.

Reply:
pwd
# /Users/kezhenxu94/workspace/skywalking-python

g rev-parse HEAD
# 6002ccc5e7463be8960304c30ecae55753f001dc

make install > /dev/null 2>&1
cat <<EOF > test.py
import multiprocessing as mp
import time

from skywalking.trace.context import get_context
from skywalking import agent, config

config.init(collector='127.0.0.1:11800', service='your awesome service')
agent.start()

def foo():
    with get_context().new_local_span('child before error'):
        pass

    time.sleep(2)  # this needed to flush send because python doesn't run atexit handlers on exit in forked children

    # import atexit
    # atexit._run_exitfuncs()

if __name__ == '__main__':
    p = mp.Process(target = foo, args = ())

    with get_context().new_local_span('parent before start'):
        pass

    p.start()

    time.sleep(1)

    with get_context().new_local_span('parent after start'):
        pass

    p.join()

    with get_context().new_local_span('parent after join'):
        pass

    time.sleep(5)
EOF

docker run -it --rm -p 11800:11800 -p 12800:12800 -d apache/skywalking-oap-server:8.6.0-es6
# 64fb2bf0e8b8030c69058d79c114969ab46207fe9391913cc91080528f510005

# Start to test
# Generate traffic
python3 test.py
# skywalking [MainThread] [WARNING] failed to install plugin sw_aiohttp
# skywalking [MainThread] [WARNING] failed to install plugin sw_django
# skywalking [MainThread] [WARNING] failed to install plugin sw_elasticsearch
# skywalking [MainThread] [WARNING] failed to install plugin sw_flask
# skywalking [MainThread] [WARNING] failed to install plugin sw_kafka
# skywalking [MainThread] [WARNING] failed to install plugin sw_psycopg2
# skywalking [MainThread] [WARNING] failed to install plugin sw_pymongo
# skywalking [MainThread] [WARNING] failed to install plugin sw_pymysql
# skywalking [MainThread] [WARNING] failed to install plugin sw_pyramid
# skywalking [MainThread] [WARNING] failed to install plugin sw_rabbitmq
# skywalking [MainThread] [WARNING] failed to install plugin sw_redis
# skywalking [MainThread] [WARNING] failed to install plugin sw_requests
# skywalking [MainThread] [WARNING] failed to install plugin sw_sanic
# skywalking [MainThread] [WARNING] failed to install plugin sw_tornado
# skywalking [MainThread] [WARNING] failed to install plugin sw_urllib3
# skywalking [MainThread] [WARNING] failed to install plugin sw_aiohttp
# skywalking [MainThread] [WARNING] failed to install plugin sw_django
# skywalking [MainThread] [WARNING] failed to install plugin sw_elasticsearch
# skywalking [MainThread] [WARNING] failed to install plugin sw_flask
# skywalking [MainThread] [WARNING] failed to install plugin sw_kafka
# skywalking [MainThread] [WARNING] failed to install plugin sw_psycopg2
# skywalking [MainThread] [WARNING] failed to install plugin sw_pymongo
# skywalking [MainThread] [WARNING] failed to install plugin sw_pymysql
# skywalking [MainThread] [WARNING] failed to install plugin sw_pyramid
# skywalking [MainThread] [WARNING] failed to install plugin sw_rabbitmq
# skywalking [MainThread] [WARNING] failed to install plugin sw_redis
# skywalking [MainThread] [WARNING] failed to install plugin sw_requests
# skywalking [MainThread] [WARNING] failed to install plugin sw_sanic
# skywalking [MainThread] [WARNING] failed to install plugin sw_tornado
# skywalking [MainThread] [WARNING] failed to install plugin sw_urllib3
# Query and see the span number

curl -s 'http://localhost:12800/graphql' \
-X 'POST' \
-H 'Content-Type: application/json;charset=utf-8' \
--data-binary '{"query":"query queryTraces($condition: TraceQueryCondition) {\n  data: queryBasicTraces(condition: $condition) { traces { endpointNames } total }}","variables":{"condition":{"queryDuration":{"start":"2021-06-23 131309","end":"2021-06-23 232809","step":"SECOND"},"traceState":"ALL","paging":{"pageNum":1,"pageSize":15,"needTotal":true},"queryOrder":"BY_START_TIME","tags":[]}}}' | jq '.data.data.total'

# 4 <================== 

python3 test.py
# skywalking [MainThread] [WARNING] failed to install plugin sw_aiohttp
# skywalking [MainThread] [WARNING] failed to install plugin sw_django
# skywalking [MainThread] [WARNING] failed to install plugin sw_elasticsearch
# skywalking [MainThread] [WARNING] failed to install plugin sw_flask
# skywalking [MainThread] [WARNING] failed to install plugin sw_kafka
# skywalking [MainThread] [WARNING] failed to install plugin sw_psycopg2
# skywalking [MainThread] [WARNING] failed to install plugin sw_pymongo
# skywalking [MainThread] [WARNING] failed to install plugin sw_pymysql
# skywalking [MainThread] [WARNING] failed to install plugin sw_pyramid
# skywalking [MainThread] [WARNING] failed to install plugin sw_rabbitmq
# skywalking [MainThread] [WARNING] failed to install plugin sw_redis
# skywalking [MainThread] [WARNING] failed to install plugin sw_requests
# skywalking [MainThread] [WARNING] failed to install plugin sw_sanic
# skywalking [MainThread] [WARNING] failed to install plugin sw_tornado
# skywalking [MainThread] [WARNING] failed to install plugin sw_urllib3
# skywalking [MainThread] [WARNING] failed to install plugin sw_aiohttp
# skywalking [MainThread] [WARNING] failed to install plugin sw_django
# skywalking [MainThread] [WARNING] failed to install plugin sw_elasticsearch
# skywalking [MainThread] [WARNING] failed to install plugin sw_flask
# skywalking [MainThread] [WARNING] failed to install plugin sw_kafka
# skywalking [MainThread] [WARNING] failed to install plugin sw_psycopg2
# skywalking [MainThread] [WARNING] failed to install plugin sw_pymongo
# skywalking [MainThread] [WARNING] failed to install plugin sw_pymysql
# skywalking [MainThread] [WARNING] failed to install plugin sw_pyramid
# skywalking [MainThread] [WARNING] failed to install plugin sw_rabbitmq
# skywalking [MainThread] [WARNING] failed to install plugin sw_redis
# skywalking [MainThread] [WARNING] failed to install plugin sw_requests
# skywalking [MainThread] [WARNING] failed to install plugin sw_sanic
# skywalking [MainThread] [WARNING] failed to install plugin sw_tornado
# skywalking [MainThread] [WARNING] failed to install plugin sw_urllib3

# Query and see the span number
curl -s 'http://localhost:12800/graphql' \
-X 'POST' \
-H 'Content-Type: application/json;charset=utf-8' \
--data-binary '{"query":"query queryTraces($condition: TraceQueryCondition) {\n  data: queryBasicTraces(condition: $condition) { traces { endpointNames } total }}","variables":{"condition":{"queryDuration":{"start":"2021-06-23 131309","end":"2021-06-23 232809","step":"SECOND"},"traceState":"ALL","paging":{"pageNum":1,"pageSize":15,"needTotal":true},"queryOrder":"BY_START_TIME","tags":[]}}}' | jq '.data.data.total'

# 8 <===============

Reply:
I do not get this result, and frankly I find it weird that you do since this version does not do anything special for grpc on fork. Tested with both python 3.8 and 3.9 on with clean build I get three spans and a deadlock. Also, I do not see `sw_celery` listed for plugins which this PR contains, are you running this PR?

Reply:
Just tried with previous master without celery or fork changes and still only get three spans, but no deadlock.

Reply:
Can you add `time.sleep(5)` at the end of the program and retry? Like what I did 👆

Reply:
Same result for me. Is possible it is something in my env. Can you try with also a span BEFORE the fork with and without a 2 sec delay to see if maybe grpc is handling the fork ok on its own?

Reply:
But also, the missing `failed to install plugin sw_celery` tells me you are not running this PR.

Reply:
> But also, the missing `failed to install plugin sw_celery` tells me you are not running this PR.

I was running on master branch, not this PR

Reply:
Same result with or without a 2 sec delay

Reply:
I have tried a few more times, with upstream/master, and still bad results. I did get one run where I got all 4 spans but the rest of the runs were 3 spans with a couple of deadlocks. Apart from that, upstream/master can not possibly run correctly in a multiprocessing scenario because on fork() no other threads are duplicated in the child (like report or heartbeat), they need to be explicitly recreated in a fork child (which I do in this PR).

I don't have time allocated now to look into the grpc issue but I do know that http protocol in this PR works with fork() for sure. So how do you want to proceed? I could remove that warning message if you want, or change it to something a little less absolute like "fork() may not work correctly with grpc protocol"? But in general this PR does not change anything about how grpc worked before, just fixes the http protocol and adds restart of report and heartbeat threads in fork() child. And also the celery plugin of course.

Reply:
BTW, this is not the end of the road though. Our internal stress tests show problems with spans mixing or disappearing so I need to go back into core functionality and fix all that, maybe overhaul how span context is tracked like in the Node agent (especially since async wansn't originally a design consideration in this agent). So treat this PR as a single step towards getting all that fixed.